### PR TITLE
RandomizedSearchCV Implementation

### PR DESCRIPTION
**Pull request by whensbrunch**
Love your work on the surprise lib! My team and I have used it extensively.

I had great results tuning hyperparameters using an approach similar to RandomizedSearchCV, so I wanted to give back to the project by implementing an analog for surprise.

This PR accomplishes the following:
- [x] creates a RandomizedSearchCV class in search.py based heavily on sklearn
- [x] adjust __init__.py so the new class can be imported
- [x] add version-related helper function to utils.py
- [x] write unit tests in test_search.py

All pytest checks are passing ✅ 

Right now there's a lot of duplicate code between GridSearchCV and RandomizedSearchCV that could be cleaned up using an abstract base class. If that's of interest to you, it's definitely something I can work on.

Cheers!

### CODE REVIEW

**Comment by NicolasHug** [comment]
That looks awesome! Thank you so much :) randomized grid search was definetely a missing feature.

A few remarks / questions:

- scipy should be added to all requirements files, including for travis: that's why the checks are failing right now.
- While we're at it, we could force scipy version to be `scipy >= 1.0`. That would avoid the need for `sp_version` and `_parse_version`. What's your opinion?

>Right now there's a lot of duplicate code between GridSearchCV and RandomizedSearchCV that could be cleaned up using an abstract base class. If that's of interest to you, it's definitely something I can work on.

Go for it! :)

**Comment by whensbrunch** [comment]
Great! I went ahead and refactored to make use of an abstract base class `BaseSearchCV`. 

> While we're at it, we could force scipy version to be scipy >= 1.0. That would avoid the need for sp_version and _parse_version. What's your opinion?

I like this. As far as I can tell, there's not a hugely important reason to support Scipy before 1.0.0. The latest commits remove `sp_version` and `_parse_version`!

File: tests/test_search.py [hunk 1]
```diff
@@ -8,17 +8,32 @@
...
 from surprise.model_selection import KFold
 from surprise.model_selection import PredefinedKFold
-from surprise.model_selection import GridSearchCV
+from surprise.model_selection import GridSearchCV, RandomizedSearchCV
 from surprise.model_selection import cross_validate
 
 
-def test_parameter_combinations():
+# Tests for GridSearchCV class
+def test_basesearchcv_parse_options():
```
Comment (NicolasHug) [inline tests/test_search.py]: Please add short docstring explening the purpose of the test

**Comment by NicolasHug** [comment]
Awesome work man, I just added minor comment (see above).

If I understand it correctly the only difference between the two subclasses is the `self.param_combinations`?

File: tests/test_search.py (hunk 1 above)
Comment (whensbrunch) [reply tests/test_search.py]: All set!

**Comment by whensbrunch** [comment]
> If I understand it correctly the only difference between the two subclasses is the self.param_combinations?

That's right! There's also an additional parameter called `n_iter` to indicate the number of parameter combinations to generate in RandomizedSearchCV.

The docs all compile and `_sample_parameters` will be ignored by sphinx by default because of the leading underscore. I did change the `model_selection.rst` a little to avoid showing the abstract base class.

Finally, as a sanity check, I ran

```python
param_distributions = {
    'n_factors': uniform(1, 100),
    'n_epochs': [20],
    'biased': [True, False],
    'lr_all': [.005],
    'reg_all': [.01],
}
rs = RandomizedSearchCV(SVD(), param_distributions, n_iter=10000)
```

and then looked at the distribution of the `n_factors` parameter in the generated combinations:

image

Everything looks as expected!

**Comment by NicolasHug** [comment]
Perfect, thank you so much! 💯

**Pull request merged**